// @hyzyla/pdfium/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PDFiumFormObject = exports.PDFiumShadingObject = exports.PDFiumImageObject = exports.PDFiumPathObject = exports.PDFiumTextObject = exports.PDFiumPage = exports.PDFiumLibrary = exports.PDFiumDocument = void 0;
var document_1 = require("./document");
Object.defineProperty(exports, "PDFiumDocument", { enumerable: true, get: function () { return document_1.PDFiumDocument; } });
var library_1 = require("./library");
Object.defineProperty(exports, "PDFiumLibrary", { enumerable: true, get: function () { return library_1.PDFiumLibrary; } });
var page_1 = require("./page");
Object.defineProperty(exports, "PDFiumPage", { enumerable: true, get: function () { return page_1.PDFiumPage; } });
var objects_1 = require("./objects");
Object.defineProperty(exports, "PDFiumTextObject", { enumerable: true, get: function () { return objects_1.PDFiumTextObject; } });
Object.defineProperty(exports, "PDFiumPathObject", { enumerable: true, get: function () { return objects_1.PDFiumPathObject; } });
Object.defineProperty(exports, "PDFiumImageObject", { enumerable: true, get: function () { return objects_1.PDFiumImageObject; } });
Object.defineProperty(exports, "PDFiumShadingObject", { enumerable: true, get: function () { return objects_1.PDFiumShadingObject; } });
Object.defineProperty(exports, "PDFiumFormObject", { enumerable: true, get: function () { return objects_1.PDFiumFormObject; } });
